File: carrera de caballos/web/js/model/Deck.js
/**
 * Mazo: draw pile y descarte.
 */
class Deck {
  constructor() {
    this.drawPile = [];
    this.discardPile = [];
    this.allIds = new Set();
  }

  addToDraw(card) {
    if (!card) throw new Error('card no puede ser nula');
    if (this.allIds.has(card.id)) throw new Error('Carta duplicada: ' + card.id);
    this.allIds.add(card.id);
    this.drawPile.push(card);
  }

  draw() {
    return this.drawPile.shift() || null;
  }

  discard(card) {
    if (card) this.discardPile.push(card);
  }

  reshuffleDiscardIntoDraw(rng) {
    if (rng && typeof rng.shuffle === 'function') {
      rng.shuffle(this.discardPile);
    } else {
      for (let i = this.discardPile.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [this.discardPile[i], this.discardPile[j]] = [this.discardPile[j], this.discardPile[i]];
      }
    }
    this.drawPile.push(...this.discardPile);
    this.discardPile.length = 0;
  }

  get isDrawEmpty() {
    return this.drawPile.length === 0;
  }

  drawSize() {
    return this.drawPile.length;
  }

  discardSize() {
    return this.discardPile.length;
  }

  hasCardsToDrawOrDiscard() {
    return this.drawPile.length > 0 || this.discardPile.length > 0;
  }
}

window.Deck = Deck;

File: carrera de caballos/src/main/java/carrera/model/Deck.java
package carrera.model;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Deque;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Mazo: draw pile (cola LIFO/FIFO según uso) y descarte.
 * Invariantes: no duplicados; los 4 ases no están en draw ni en pista.
 */
public class Deck {
    private final Deque<Card> drawPile;
    private final List<Card> discardPile;
    private final Set<String> allIds;

    public Deck() {
        this.drawPile = new ArrayDeque<>();
        this.discardPile = new ArrayList<>();
        this.allIds = new HashSet<>();
    }

    /** Añade una carta al mazo de robo. Restricción: id único. */
    public void addToDraw(Card card) {
        if (card == null) throw new IllegalArgumentException("card no puede ser nula");
        if (!allIds.add(card.getId())) throw new IllegalArgumentException("Carta duplicada: " + card.getId());
        drawPile.addLast(card);
    }

    /** Roba una carta del mazo y la devuelve; si está vacío devuelve null. */
    public Card draw() {
        return drawPile.pollFirst();
    }

    /** Añade una carta al descarte (para rebarajar después). */
    public void discard(Card card) {
        if (card != null) discardPile.add(card);
    }

    /** Rebaraja el descarte en el mazo de robo (las cartas de pista NO se rebarajan). */
    public void reshuffleDiscardIntoDraw(java.util.Random rng) {
        if (rng != null) Collections.shuffle(discardPile, rng);
        else Collections.shuffle(discardPile);
        for (Card c : discardPile) drawPile.addLast(c);
        discardPile.clear();
    }

    public boolean isDrawEmpty() {
        return drawPile.isEmpty();
    }

    public int drawSize() {
        return drawPile.size();
    }

    public int discardSize() {
        return discardPile.size();
    }

    public boolean hasCardsToDrawOrDiscard() {
        return !drawPile.isEmpty() || !discardPile.isEmpty();
    }
}
